akka-http-core/src/main/scala/akka/http/impl/model/parser/SimpleHeaders.scala

```diff
@@ -95,7 +95,15 @@ private[parser] trait SimpleHeaders { this: Parser with CommonRules with CommonA
 
   // https://tools.ietf.org/html/rfc6265#section-4.2
   def `cookie` = rule {
-    oneOrMore(`cookie-pair`).separatedBy((ch(';') | ch(',')) ~ OWS) ~ EOI ~> (`Cookie`(_))
+    oneOrMore(`optional-cookie-pair`).separatedBy(';' ~ OWS) ~ EOI ~> { pairs ⇒
+      val validPairs = pairs.collect { case Some(p) ⇒ p }
+      `Cookie` {
+        if (validPairs.nonEmpty) validPairs
+        // Parsing infrastructure requires to return an HttpHeader value here but it is not possible
+        // to create a Cookie header without elements, so we create a dummy instance.
+        else HttpCookiePair("__dummy__", "__No_more_cookies_left_after_discarding_invalid_ones__") :: Nil
```

Comment:
I agree that throwing an exception is probably the way to go.
Basically what we want to do here is "fail late", i.e. abort parsing into a ModeledHeader after having already decided that we (probably) can.
Actually it might even suffice to simply `throw HeaderParser.RuleNotFoundException`, which should fall back to creating a `RawHeader`. Could you try that, Johannes?